src/lookup.h

```diff
@@ -0,0 +1,267 @@
+            mi->_type = FRAME_KERNEL;
+        } else {
+            mi->_name = _symbols->indexOf(name);
+            mi->_sig = _symbols->indexOf("()L;");
+            mi->_type = FRAME_NATIVE;
+        }
+    }
+
+    bool fillJavaMethodInfo(MethodInfo* mi, jmethodID method, bool first_time) {
+        if (VMMethod::isStaleMethodId(method)) {
+            return false;
+        }
+
+        jclass method_class = NULL;
+        char* class_name = NULL;
+        char* method_name = NULL;
+        char* method_sig = NULL;
+        char* source_name = NULL;
+
+        jvmtiEnv* jvmti = VM::jvmti();
+        jvmtiError err;
+
+        if ((err = jvmti->GetMethodName(method, &method_name, &method_sig, NULL)) == 0 &&
+            (err = jvmti->GetMethodDeclaringClass(method, &method_class)) == 0 &&
+            (err = jvmti->GetClassSignature(method_class, &class_name, NULL)) == 0 &&
+            (err = jvmti->GetSourceFileName(method_class, &source_name)) == 0) {
+            mi->_class = _classes->lookup(class_name + 1, strlen(class_name) - 2);
+            mi->_file = _classes->lookup(source_name);
+            mi->_name = _symbols->indexOf(method_name);
+            mi->_sig = _symbols->indexOf(method_sig);
+        }
+
+        if (method_class) {
+            _jni->DeleteLocalRef(method_class);
+        }
+        jvmti->Deallocate((unsigned char*)method_sig);
+        jvmti->Deallocate((unsigned char*)method_name);
+        jvmti->Deallocate((unsigned char*)class_name);
+        jvmti->Deallocate((unsigned char*)source_name);
+
+        if (err != 0) {
+            return false;
+        }
+
+        if (first_time && jvmti->GetMethodModifiers(method, &mi->_modifiers) != 0) {
+            mi->_modifiers = 0;
+        }
+
+        if (first_time && jvmti->GetLineNumberTable(method, &mi->_line_number_table_size, &mi->_line_number_table) != 0) {
+            mi->_line_number_table_size = 0;
+            mi->_line_number_table = NULL;
+        }
+
+        mi->_type = FRAME_INTERPRETED;
+        return true;
+    }
+
+    void fillJavaClassInfo(MethodInfo* mi, u32 class_id) {
+        mi->_class = class_id;
+        mi->_name = _symbols->indexOf("");
+        mi->_sig = _symbols->indexOf("()L;");
+        mi->_modifiers = 0;
+        mi->_line_number_table_size = 0;
+        mi->_line_number_table = NULL;
+        mi->_type = FRAME_INLINED;
+    }
+
+  public:
+    Lookup(MethodMap* method_map, Dictionary* classes) : Lookup(method_map, classes, nullptr, nullptr) {}
+
+    Lookup(MethodMap* method_map, Dictionary* classes, Index* packages, Index* symbols) :
+        _method_map(method_map),
+        _classes(classes),
+        _packages(packages),
+        _symbols(symbols),
+        _jni(VM::jni()) {
+
+        if (_packages == nullptr) {
+            _owns_packages = true;
+            _packages = new Index();
+        }
+        if (_symbols == nullptr) {
+            _owns_symbols = true;
+            _symbols = new Index();
+        }
```

Comment:
This "ownership" concept does not seem the best design solution to me.
OTLP format does not use packages, IIRC?

Reply:
I agree it's a bit awkward, fixed in e1b90df72a7291d24a8103783d2f8d9308557339 and 4eef22b649c7b98b9d5b383f4c927bf9debd8390